File: top_interview_questions_easy/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
      stack<char> charStack;
      for(char charC : s) {
        if(charC == '(') {
          charStack.push(')');
        }else if(charC == '[') {
          charStack.push(']');
        }else if(charC == '{') {
          charStack.push('}');
        }else {
          if(charStack.empty() || charStack.top() != charC) {
            return false;
          }
          // 괄호 짝이 맞으면 charStack을 비운다.
          charStack.pop();
        }
      }
      // charStack이 비워지면 괄호 짝이 맞는 것이 되므로 true
      return charStack.empty();
    }
};

File: top_interview_questions_easy/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False

        dict = {")": "(",
                "}": "{",
                "]": "["}
        stack = []
        for n in s:
            if n not in dict:
                # 열린 괄호인 경우 stack에 넣음
                stack.append(n)
            # 닫힌 괄호인 경우 가장 최근 아이템과 비교해 같지 않을 경우 거짓
            elif (not stack) or (dict[n] != stack.pop()):
                return False
        return len(stack) == 0





